src/i_joystick.c

```diff
@@ -76,6 +81,246 @@ static int joystick_physical_buttons[NUM_VIRTUAL_BUTTONS] = {
+        SDL_QuitSubSystem(SDL_INIT_GAMECONTROLLER);
+    }
+}
+
+static int FindFirstGamepad(void)
+{
+    int i;
+    int gamepadindex = -1;
+
+    for (i = 0; i < SDL_NumJoysticks(); ++i)
+    {
+        if (SDL_IsGameController(i))
+        {
+            gamepadindex = i;
+            break;
+        }
+    }
+
+    return gamepadindex;
+}
+
+static int FindSpecificGamepad(SDL_JoystickGUID guid)
+{
+    SDL_JoystickGUID dev_guid;
+    int i;
+    int gamepadindex = -1;
+
+    for (i = 0; i < SDL_NumJoysticks(); ++i)
+    {
+        dev_guid = SDL_JoystickGetDeviceGUID(i);
+        if (!memcmp(&guid, &dev_guid, sizeof(SDL_JoystickGUID)))
+        {
+            gamepadindex = i;
+            break;
+        }
+    }
+
+    return gamepadindex;
+}
+
+static int DeviceIndexGamepad(void)
+{
+    SDL_JoystickGUID guid, dev_guid;
+    int index = -1;
+
+    if (strcmp(joystick_guid, ""))
+    {
+        guid = SDL_JoystickGetGUIDFromString(joystick_guid);
+
+        // First, look for the gamepad at the previously-used index.
+        if (joystick_index >= 0 && joystick_index < SDL_NumJoysticks())
+        {
+            dev_guid = SDL_JoystickGetDeviceGUID(joystick_index);
+            if (!memcmp(&guid, &dev_guid, sizeof(SDL_JoystickGUID)))
+            {
+                return joystick_index;
+            }
+        }
+
+        // Maybe the index has moved?
+        index = FindSpecificGamepad(guid);
+    }
+
+    // Find a gamepad for the first time. Or, if the previous gamepad isn't
+    // present, see if a different one is available.
+    if (index < 0)
+    {
+        index = FindFirstGamepad();
+    }
+
+    return index;
+}
+
+void I_InitGamepad(void)
+{
+    SDL_JoystickGUID guid;
+    int index;
+
+    if (!use_gamepad)
+    {
+        return;
+    }
+
+    if (SDL_Init(SDL_INIT_GAMECONTROLLER) < 0)
+    {
+        return;
+    }
+
+    index = DeviceIndexGamepad();
+
+    if (index < 0)
+    {
+        printf("I_InitGamepad: No gamepad found.\n");
+        SDL_QuitSubSystem(SDL_INIT_GAMECONTROLLER);
+        return;
+    }
+
+    gamepad = SDL_GameControllerOpen(index);
+
+    if (gamepad == NULL)
+    {
+        printf("I_InitGamepad: Failed to open gamepad.\n");
+        SDL_QuitSubSystem(SDL_INIT_GAMECONTROLLER);
+        return;
+    }
+
+    joystick_index = index;
+
+    if (strcmp(joystick_guid, ""))
+    {
+        joystick_guid = malloc(33);
+    }
+
+    guid = SDL_JoystickGetDeviceGUID(joystick_index);
+    SDL_JoystickGetGUIDString(guid, joystick_guid, 33);
+
+    // GameController events do not fire if Joystick events are disabled.
+    SDL_JoystickEventState(SDL_ENABLE);
+    SDL_GameControllerEventState(SDL_ENABLE);
+
+    printf("I_InitGamepad: %s\n", SDL_GameControllerName(gamepad));
+    I_AtExit(I_ShutdownGamepad, true);
+}
+
+static int GetTriggerStateGamepad(SDL_GameControllerAxis trigger)
+{
+    int value;
+
+    value = SDL_GameControllerGetAxis(gamepad, trigger);
+
+    return (value > TRIGGER_THRESHOLD);
```

Comment:
Should this be configurable?

Reply:
I'm guessing that hardcoded constant is fine for this. I don't see this offered as an option in other sourceports, nor in other games in general.